contracts/interfaces/slash-indicator/ISlashFastFinality.sol

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.9;
+
+import "./IBaseSlash.sol";
+
+interface ISlashFastFinality is IBaseSlash {
+  /**
+   * @dev Emitted when the configs to slash fast finality is updated. See the method `getFastFinalitySlashingConfigs`
+   * for param details.
+   */
+  event FastFinalitySlashingConfigsUpdated(uint256 slashFastFinalityAmount, uint256 fastFinalityJailUntilBlock);
+
+  /**
+   * @dev Slashes for fast finality.
+   *
+   * Requirements:
+   * - Only whitelisted addresses are allowed to call.
+   *
+   * Emits the event `Slashed` if the fast finality evidence of the two headers valid.
+   */
+  function slashFastFinality(
+    address consensusAddr,
+    bytes memory voterPublicKey,
+    uint256 targetBlockNumber,
+    bytes32[2] memory targetBlockHash,
+    bytes[][2] memory listOfPublicKey,
+    bytes[2] memory aggregatedSignature
```

Comment:
Since Calldata has already been paid for with the transaction, if you don’t modify a parameter to a function, then don’t bother copying the function to memory and just read the value from calldata.